jni/src/faiss_wrapper.cpp

```diff
@@ -682,6 +806,15 @@ void InternalTrainIndex(faiss::Index * index, faiss::idx_t n, const float* x) {
     }
 }
 
+void InternalTrainBinaryIndex(faiss::IndexBinary * index, faiss::idx_t n, const float* x) {
+    if (auto * indexIvf = dynamic_cast<faiss::IndexBinaryIVF*>(index)) {
+        indexIvf->make_direct_map();
+    }
+    if (!index->is_trained) {
+        index->train(n, reinterpret_cast<const uint8_t*>(x));
```

Comment:
I don't think this reinterpret will work correctly. You cannot simply reinterpret `float*` to `uint8_t*`. The reinterpreted value in uint8_t will be something meaningless. Please check.

It need to be converted from float array to byte array through loop here. However this will use more memory for binary vector training compared to float vector training. Therefore, we need to store the training vector as byte from the start instead of float.

Reply:
Thanks @heemin32. I updated PR to support passing byte array at start in [TrainingDataConsumer.java](https://github.com/opensearch-project/k-NN/compare/800c7c565a32419b4bcc539c83ca15e682fdc2f7..7294db4988d1d2e6111e4d4fbc401e77e57ce7b9#diff-6a607059c7d3f059ac5bf61ac1947bf96c4cdb7182b114a910cb9b4bb3f6a8dbR44-R68).

Also updated freeVectorData method to support binary index deletion.